### ISSUE TITLE
When linking IdPs to organizations, wrong IdPs show up after clicking "back" and "forward" in the browser.

### ISSUE THREAD

**Issue by DGuhr** [keycloak/keycloak#33269]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

organizations

### Describe the bug

Topic is the new organization feature. I found that the IDP linking has a bug currently where IDPs that are not linked to an organization are shown in step 2 of the identity-first-flow under some circumstances. 

### Version

keycloak nightly (pulled 25.09.2024)

### Regression

- [ ] The issue is a regression

### Expected behavior

Only the IdPs linked to the respective org show up, no matter the circumstances

### Actual behavior

IdPs present in the realm but not linked to the org show up.

### How to Reproduce?

**Steps to reproduce:**

1.  Run keycloak nightly with org feature enabled: `docker run --name kc-orgs -d -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin -p 8080:8080 quay.io/keycloak/keycloak start-dev --features organization`
2. Go to the admin ui, create a realm "kcorgs", enable the organization feature in realm settings  and create 2 organizations foo with domain [foo.com](http://foo.com/) and bar with  [bar.com](http://bar.com/)
3. Create 2 realm users, foo / [foo@foo.com](mailto:foo@foo.com) and bar / [bar@bar.com](mailto:bar@bar.com)
5. Add the users as member to the respective org (foo to foo, bar to bar, so each org has one member)
6. Create two mocked identity providers, e.g. google and github (I just added some fake credentials, nothing else, so they won't "really" work - just to see their respective button on the login page)
7. Link the IdPs to the orgs so each org has one linked IdP:  foo -> github, bar -> google - here I set the domain respectively to foo.com for the foo org and bar.com for the bar org. (Not entirely sure this has to be set up), else default settings.
8. Now open http://localhost:8080/realms/kcorgs/account and use "foo@foo.com"
9. On the next page, the github IdP is correctly shown. Now click the "back" button in the browser. Then click the "Forward" button in the browser.
Result: Both, Google and GitHub IdPs are shown, though foo has only github linked.

Video showing Setup and Problem: https://youtu.be/5EIJL2v5kCw 

### Anything else?

Using chrome Version 128.0.6613.138 on MacOS here

On a side note, when I use the nightly and add an alias to the IdPs, such as foo-github and bar-google, the icons are not shown correctly on the login page, just boxes.

**Comment by sguilhen** [comment]
@DGuhr I've tried to reproduce and I think I was partially able to do it. When I hit "Back" on the browser, the identity-first page is empty, so when I hit "Forward" again there's no user set and all public IDPs (including ones belonging to orgs) are shown. This is expected behavior. The only thing I noticed is that after I hit "Forward" the username is still there (not empty like the first screen).

**Comment by sguilhen** [comment]
This part "On a side note, when I use the nightly and add an alias to the IdPs, such as foo-github and bar-google, the icons are not shown correctly on the login page, just boxes." - I was able to reproduce too.. if the default alias is changed, the icons don't work correctly.

@ssilvert @edewit - think we have an issue with icons when the alias is not the default one.

**Comment by tahvane1** [comment]
I'm also confirming this not able to show idp login link ONLY on organization level

**Comment by sguilhen** [comment]
> So the actual bug is that the field should be reset, right?

Yeah, that is what I think too.

> I just tried to set "hide on loginpage" at the realm level under Identity Providers, and switch it on at the org-level linked provider, but it seems both are internally the same, currently, which prevents a way to say "only show this provider when orgA tries to log in, not on the generic realm level".

So if I got this right, we now can either have the IDP available in login pages or not, a behavior controlled by the `hideOnLogin` setting, but we don't have a way to say "this is available for login ONLY when in the context of the org". You might not want to make the IDP unavailable for login screen (which is what the `hideOnLogin` flag does), but rather make it available only when we detect that the user is member of the org.

I don't recall exactly all the reasoning behind the decision to make all public IDPs available by default, would need to revisit this with @pedroigor next week, but this is set here:

https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java#L485-L487

This is the same as calling the constructor with `(idpBean, false, false)` - second param indicates if we should have only org brokers, third param indicates if we should have only realm brokers. Both being false is the equivalent of saying "give me all public enabled idps, both realm-level ones and org-linked ones."

Changing this block to build the org-aware bean with `(idpBean, false, true)` would by default fetch only the realm-level idps, all org idps would be hidden.

This breaks a bunch of tests, which is why I would need to revisit the reasons for the assumption that this is a good default in the first place.

**Comment by DGuhr** [comment]
You are correct, afaict. I didn't look at the codelevel though, as I am on the phone currently. 

One thing I am chewing in my hand is that this is a clash of "iam boundary" definitions. If you define your "iam boundary"* to be a "realm", which Keycloak did up until now, it makes perfect sense to show all available IdPs that are not hidden on login. Hence the current default.

BUT:
If you now break up this long-standing definition, and say "the iam boundary is an org inside a realm", this shifts the boundary to something smaller than realm-level, which is here called organization. Now with this switch, prior-sane defaults may not make sense anymore, as illustrated in the above use case. 

I think all I want to say here is that there might bw two defaults needed:
- Realm without Orgs: Use the default "as is", possibility to login qith associated IdPs is the default, unless hidden.
- Realm with orgs enabled: Revert the default due to a smaller iam boundary, the org. 

And my head still explores a possible option c or d, but I cannot put into words at the moment, because it is more a vague feeling that the above is suboptimal, as it would lead to more complexity when developing basically any feature in keycloak. What I mean here is: Now one does not only need to think of "What is a sane and secure default for my feature foo on realm-level", but also "and does it also make sense on org-level or do I need extra code to switch it here".

This got a bit philosophical, nevertheless relevant I guess. I'd be happy to get links to the proposals, rfc's etc. as I am a bit late to the party and trust that there was some thought put into exactly these kinds of scenarios at the border of the iam boundary. :)

*I came up with this term right now, lacking a better one. I hope it transports what I mean.

**Comment by sguilhen** [comment]
Yeah, In general I agree with your assessment here about the shifting boundaries. I would like to resume this discussion next week when @pedroigor is back from his PTO.

**Comment by pedroigor** [comment]
@DGuhr Glad to see you around :)

As mentioned by @sguilhen, we are opinionated about certain behaviors and how brokers are shown in login pages is one of these cases. Right now, org brokers are shown when:

* The email matches an organization so that only the organization brokers are shown
* The user is a realm user or does not exist yet and we show both realm and public org brokers

The original issue you described is something we can improve and see how to properly handle the back button.

**Comment by hhao01-becls** [comment]
> @DGuhr Glad to see you around :)
> 
> As mentioned by @sguilhen, we are opinionated about certain behaviors and how brokers are shown in login pages is one of these cases. Right now, org brokers are shown when:
> 
> * The email matches an organization so that only the organization brokers are shown
> * The user is a realm user or does not exist yet and we show both realm and public org brokers
> 
> The original issue you described is something we can improve and see how to properly handle the back button.

If the org broker binds a domain and a user's email does not match the domain of an organization, when the user try to sign in, I think the org broker should not be shown, since I think this is a risk to expose the other organization's IDP link to the users don't belong to the organization.

**Comment by sguilhen** [comment]
This is likely to be handled by https://github.com/keycloak/keycloak/issues/36309
